---
env---
title: "Analysis Notebook"
output:
  html_notebook
---

# Loading Libraries

```{r}
library(ArchR)
library(Seurat)
library(Dune)

BiocManager::install("BSgenome.Hsapiens.UCSC.hg38")
BiocManager::install("Dune")
# library("BSgenome.Hsapiens.UCSC.hg38")

addArchRGenome("hg38")
```

## Create Arrow Object

```{r}
inputFiles <- "data/pbmc_unsorted_10k_atac_fragments.tsv.gz"
names(inputFiles) <- "PBMC_10k"

ArrowFiles <- createArrowFiles(
  inputFiles = inputFiles,
  minTSS = 0,
  minFrags = 500, 
  addTileMat = TRUE,
  addGeneScoreMat = TRUE
  #force = TRUE
)
```

```{r}
proj <- ArchRProject(ArrowFiles)
```

```{r}
# saveArchRProject(ArchRProj = proj, outputDirectory = "clean-save", load = FALSE)
```

# Importing RNA Data

```{r}
library(ggplot2)
library(gridExtra)  # for grid.arrange function

# Function to create an individual violin plot for a given metric
createViolinPlot <- function(df, metric, violinColor="skyblue") {
  ggplot(df, aes_string(x = "factor('Metrics')", y = metric)) + 
    geom_violin(fill = violinColor) +
    geom_jitter(width = 0.4, size = 0.5, alpha = 0.25) +
    labs(y = metric) + 
    theme_minimal() + 
    theme(axis.title.x=element_blank())
}

plotMetricsSideBySide <- function(proj, violinColor="skyblue") {
  df <- data.frame(
    UMIs = proj$Gex_nUMI,
    Genes = proj$Gex_nGenes,
    MitoRatio = proj$Gex_MitoRatio
  )

  # Create individual plots for each metric
  plotUMIs <- createViolinPlot(df, "UMIs", violinColor)
  plotGenes <- createViolinPlot(df, "Genes", violinColor)
  plotMito <- createViolinPlot(df, "MitoRatio", violinColor)

  # Combine plots side by side
  grid.arrange(plotUMIs, plotGenes, plotMito, ncol=3)
}
```

```{r clustering-confusion-matrix}
library(caret)
library(pheatmap)

plot_clusters_confusion_matrix <- function(proj) {
  cluster_names <- grep("^Clusters", names(proj), value = TRUE)
  
  for (i in seq_along(cluster_names)) {
    cluster_name <- cluster_names[i]
    
    cM <- confusionMatrix(paste0(proj$cluster_name), paste0(proj$Sample))
    cM <- cM / Matrix::rowSums(cM)
    
    p <- pheatmap(
      mat = as.matrix(cM),
      color = paletteContinuous("whiteBlue"),
      border_color = "black",
      main = cluster_name
    )
    
    # Save or display the plot as needed
    # For example, to save the plot:
    # png(paste0(cluster_name, "_pheatmap.png"))
    return(p)
    # dev.off()
  }
}
```

## RNA QC

```{r}
pbmc.data <- Read10X_h5("data/pbmc_unsorted_10k_filtered_feature_bc_matrix.h5", unique.features = TRUE)

pbmc <- CreateSeuratObject(counts = pbmc.data$`Gene Expression`, project = "pbmc10k")
```

```{r}
pbmc[["percent.mt"]] <- PercentageFeatureSet(pbmc, pattern = "^MT-")

VlnPlot(pbmc, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
```

```{r}
unique_feature_threshold_min <- 600
unique_feature_threshold_max <- 3100

percent_mito_threshold <- 19

pbmc <- subset(pbmc, subset = nFeature_RNA > unique_feature_threshold_min
               & nFeature_RNA < unique_feature_threshold_max
               & percent.mt < percent_mito_threshold)

VlnPlot(pbmc, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
```

```{r}
rm(pbmc)
rm(pbmc.data)
```

```{r}
# saveRDS(pbmc, file = "/home/ferry/Documents/University/end_of_summer_project/saves/qc_pbmc.rds")
```

## Integrating RNA

# ```{r}
seRNA <- import10xFeatureMatrix(
    input = c("data/pbmc_unsorted_10k_filtered_feature_bc_matrix.h5"),
    names = c("PBMC_10k")
)

#seRNA <- readRDS("/home/ferry/Documents/University/end_of_summer_project/saves/qc_pbmc.rds")

proj <- addGeneExpressionMatrix(input = proj, seRNA = seRNA, force = TRUE)
```

```{r}
# saveArchRProject(ArchRProj = proj, outputDirectory = "atac-rna-clean", load = FALSE)
```

```{r}
proj <- loadArchRProject(path = "atac-rna-clean")
```

```{r}
plotMetricsSideBySide(proj)
```

```{R}
minFeatures <- 600
maxFeatures <- 4000
maxMT <- 0.019

# Filter out cells with NA values for the relevant metrics
proj <- proj[!is.na(proj$Gex_nGenes) & 
             !is.na(proj$Gex_MitoRatio)]

proj <- proj[proj$Gex_nGenes > minFeatures
  & proj$Gex_nGenes < maxFeatures
  & proj$Gex_MitoRatio < maxMT]
```

```{r}
plotMetricsSideBySide(proj)
```

# ATAC QC

## Plotting

```{r}
df <- getCellColData(proj, select = c("log10(nFrags)", "TSSEnrichment"))
df

p <- ggPoint(
    x = df[,1], 
    y = df[,2], 
    colorDensity = TRUE,
    continuousSet = "sambaNight",
    xlabel = "Log10 Unique Fragments",
    ylabel = "TSS Enrichment",
    xlim = c(log10(500), quantile(df[,1], probs = 0.99)),
    ylim = c(0, quantile(df[,2], probs = 0.99))
) + geom_hline(yintercept = 4, lty = "dashed") + geom_vline(xintercept = 3, lty = "dashed")

plotPDF(p, name = "TSS-vs-Frags.pdf", ArchRProj = proj, addDOC = FALSE)

p
```

## Filtering

```{r}
# For now, I want to keep the low TSS and unique fragment cells in the analysis, just because there are so many. I understand that quantity !> quality.
minTSS <- 12
minFrags <- 1200

proj <- proj[proj$TSSEnrichment > minTSS & proj$nFrags > minFrags]
```

```{r}
df <- getCellColData(proj, select = c("log10(nFrags)", "TSSEnrichment"))
df

p <- ggPoint(
    x = df[,1], 
    y = df[,2], 
    colorDensity = TRUE,
    continuousSet = "sambaNight",
    xlabel = "Log10 Unique Fragments",
    ylabel = "TSS Enrichment",
    xlim = c(log10(500), quantile(df[,1], probs = 0.99)),
    ylim = c(0, quantile(df[,2], probs = 0.99))
) + geom_hline(yintercept = 12, lty = "dashed") + geom_vline(xintercept = 3.05, lty = "dashed")

plotPDF(p, name = "TSS-vs-Frags.pdf", ArchRProj = proj, addDOC = FALSE)

p
```

```{r}
saveArchRProject(ArchRProj = proj, outputDirectory = "atac-rna-quality-controlled", load = FALSE)
```

# Dimensionality Reduciton

## ATAC LSI

```{r}
proj <- addIterativeLSI(
    ArchRProj = proj, 
    clusterParams = list(
      resolution = 0.2, 
      sampleCells = 10000,
      n.start = 10
    ),
    saveIterations = FALSE,
    useMatrix = "TileMatrix", 
    depthCol = "nFrags",
    name = "LSI_ATAC"
)
```

## RNA LSI

```{r}
proj <- addIterativeLSI(
    ArchRProj = proj, 
    clusterParams = list(
      resolution = 0.2, 
      sampleCells = 10000,
      n.start = 10
    ),
    saveIterations = FALSE,
    useMatrix = "GeneExpressionMatrix", 
    depthCol = "Gex_nUMI",
    varFeatures = 2500,
    firstSelection = "variable",
    binarize = FALSE,
    name = "LSI_RNA"
)
```

## Combining

```{r}
proj <- addCombinedDims(proj, reducedDims = c("LSI_ATAC", "LSI_RNA"), name =  "LSI_Combined")
```

# Clustering

## ATAC Clustering

```{r}
proj <- addUMAP(proj, reducedDims = "LSI_ATAC", name = "UMAP_ATAC", minDist = 0.8, force = TRUE)
```

## RNA Clustering

```{r}
proj <- addUMAP(proj, reducedDims = "LSI_RNA", name = "UMAP_RNA", minDist = 0.8, force = TRUE)
```

## Combining

```{r}
proj <- addUMAP(proj, reducedDims = "LSI_Combined", name = "UMAP_Combined", minDist = 0.8, force = TRUE)

proj <- addClusters(proj, reducedDims = "LSI_Combined", name = "Clusters", resolution = 0.4, force = TRUE)
```

## UMAP Plot

```{r}
p1 <- plotEmbedding(proj, name = "Clusters", embedding = "UMAP_ATAC", size = 1.5, labelAsFactors=F, labelMeans=F)

p2 <- plotEmbedding(proj, name = "Clusters", embedding = "UMAP_RNA", size = 1.5, labelAsFactors=F, labelMeans=F)

p3 <- plotEmbedding(proj, name = "Clusters", embedding = "UMAP_Combined", size = 1.5, labelAsFactors=F, labelMeans=F)

p1
p2
p3
```

## Cluster Evaluation

```{r}
# Get markers for RNA
markersRNA <- getMarkerFeatures(
  ArchRProj = proj,
  useMatrix = "GeneExpressionMatrix",
  groupBy = "Clusters",
  bias = c("Gex_nUMI", "Gex_nGenes"), 
  testMethod = "wilcoxon"
)

dfMarkersRNA <- as.data.frame(assay(markersRNA, "Log2FC"))

# Assuming that the columns of dfMarkersRNA are the clusters and rows are genes, 
# you can get the top N genes by absolute Log2FC
N <- 50
topNgenes <- rownames(dfMarkersRNA)[order(-abs(rowMeans(dfMarkersRNA, na.rm=TRUE)))[1:N]]

# Subset the markersRNA for only these top genes
topNMarkersRNA <- markersRNA[topNgenes, ]

plotRNA <- plotMarkerHeatmap(
  seMarker = topNMarkersRNA,
  cutOff = "FDR <= 0.01 & abs(Log2FC) >= 1", 
  transpose = TRUE
)

plotRNA

# Get markers for ATAC
# markersATAC <- getMarkerFeatures(
#   ArchRProj = proj,
#   useMatrix = "TileMatrix",
#   groupBy = "Clusters",
#   bias = c("nFrags"), 
#   testMethod = "wilcoxon"
# )
# 
# plotATAC <- plotMarkerHeatmap(
#   seMarker = markersATAC,
#   cutOff = "FDR <= 0.01 & abs(log2FC) >= 1", 
#   transpose = TRUE,
#   returnPlot = TRUE
# )

# Extract clusters and sample data into temporary vectors
tempClusters <- as.factor(proj$Clusters)
tempSample <- as.factor(proj$Sample)

# Check levels
levels(tempClusters)
levels(tempSample)

# If they don't have the same levels, you can set them to have the same levels
all_levels <- union(levels(tempClusters), levels(tempSample))
tempClusters <- factor(tempClusters, levels = all_levels)
tempSample <- factor(tempSample, levels = all_levels)

# Now compute the confusion matrix
cM <- confusionMatrix(tempClusters, tempSample)

cM <- cM / Matrix::rowSums(cM)

p <- pheatmap(
  mat = as.matrix(cM),
  color = paletteContinuous("whiteBlue"),
  border_color = "black",
  main = cluster_name
)

p
```

# Automated Clustering

```{r}
proj <- loadArchRProject(path = "atac-rna-quality-controlled")
```

```{r}
elapsedTime <- function(difference) {
  total_secs <- as.numeric(difference, units = "secs")
  hours <- as.integer(total_secs %/% 3600)
  minutes <- as.integer((total_secs %% 3600) %/% 60)
  seconds <- as.integer(total_secs %% 60)
  
  sprintf("%02d:%02d:%02d", hours, minutes, seconds)
}
```

```{r}
appendToCSV <- function(newColumnDF, filename) {
  # Check if the input is a one-dimensional dataframe
  if(ncol(newColumnDF) != 1) {
    stop("Input dataframe must be one-dimensional!")
  }
  
  # Check if the file exists
  if(!file.exists(filename)) {
    stop(paste("File", filename, "does not exist!"))
  }
  
  # Read the existing CSV file
  existingDF <- read.csv(filename, stringsAsFactors = FALSE)
  
  # Check if the number of rows matches
  if(nrow(existingDF) != nrow(newColumnDF)) {
    stop("Number of rows in the new column does not match the existing CSV!")
  }
  
  # Add the new column to the existing dataframe
  colName <- colnames(newColumnDF)
  existingDF[[colName]] <- newColumnDF[[colName]]
  
  # Write the updated dataframe back to the CSV file
  write.csv(existingDF, file = filename, row.names = FALSE)
}
```

```{r}
testClusters <- function(proj, resolutions_ATAC, resolutions_RNA,
                         iterationsList, dimsList, verbosity = 1,
                         writeLogs = TRUE,
                         logFile = "automatedClusteringLogfile.log") {
  
  if(verbosity >= 1) {
    start_time <- Sys.time()
    cat(
      strftime(start_time, format="%Y-%m-%d %H:%M:%S"),
      "- Starting Clustering Loop."
      )
  }
  
  # Check if all input lists are of the same length
  lengths <- c(length(resolutions_ATAC), length(resolutions_RNA), length(iterationsList), length(dimsList))
  
  if(length(unique(lengths)) != 1){
    stop("All input lists must be of the same length!")
  }
  
  first_col <- data.frame(1:nCells(proj))
  colnames(first_col) <- "Cell Index"
  
  # Write the dataframe to a CSV file
  write.csv(first_col, file = "ClusterAssignments.csv", row.names = FALSE)
  
  plots <- list()
  clusterAssignments <- data.frame(cellBarcode = getCellNames(proj))
  
  for(i in 1:length(resolutions_ATAC)){
    
    if(verbosity >= 1){
      cat(
        strftime(Sys.time(), format="%Y-%m-%d %H:%M:%S"),
        "- Starting iteration", i, 
        "of", length(resolutions_ATAC), "\n")
    }
    
    res_ATAC <- resolutions_ATAC[i]
    res_RNA <- resolutions_RNA[i]
    iters <- iterationsList[i]
    dims <- dimsList[i]
    
    ATAC_LSI_name <- paste0("LSI_ATAC_", i, "_Res_", gsub("\\.", "-", as.character(res_ATAC)), "_Iters_", iters, "_Dims_1to", dims)
    
    if(verbosity >= 2){
      cat(strftime(Sys.time(), format="%Y-%m-%d %H:%M:%S"),
          "- Performing dimensionality reduction and clustering for ATAC with \n",
          "Resolution =", res_ATAC, "\n",
          "Iterations = ", iters, "\n",
          "Dimensions = 1 :", dims, "\n",
          "Name = ", ATAC_LSI_name, "\n")
    }
    
    if (!writeLogs) {
      sink(logFile, append=TRUE)
    }
    
    # Dimensionality reduction and clustering for ATAC
    proj <- addIterativeLSI(
      ArchRProj = proj,
      iterations = iters,
      clusterParams = list(resolution = res_ATAC, sampleCells = 10000, n.start = 10),
      saveIterations = FALSE,
      useMatrix = "TileMatrix",
      depthCol = "nFrags",
      dimsToUse = 1:dims,
      name = ATAC_LSI_name,
      force = TRUE
    )
    
    if (!writeLogs) {
      sink()
    }
    
    RNA_LSI_name <- paste0("LSI_RNA_", i, "_Res_", gsub("\\.", "-", as.character(res_RNA)), "_Iters_", iters, "_Dims_1to", dims)
    
    if(verbosity >= 2){
      cat(strftime(Sys.time(), format="%Y-%m-%d %H:%M:%S"),
          "- Performing dimensionality reduction and clustering for RNA with:",
          "Resolution =", res_RNA,
          "Iterations = ", iters, "\n",
          "Dimensions = 1 :", dims, "\n",
          "Name = ", RNA_LSI_name, "\n")
    }
    
    if (!writeLogs) {
      sink(logFile, append=TRUE)
    }
    
    # Dimensionality reduction and clustering for RNA
    proj <- addIterativeLSI(
      ArchRProj = proj,
      iterations = iters,
      clusterParams = list(resolution = res_RNA, sampleCells = 10000, n.start = 10),
      saveIterations = FALSE,
      useMatrix = "GeneExpressionMatrix", 
      depthCol = "Gex_nUMI",
      varFeatures = 2500,
      firstSelection = "variable",
      binarize = FALSE,
      dimsToUse = 1:dims,
      name = RNA_LSI_name,
      force = TRUE
    )
    
    if (!writeLogs) {
      sink()
    }
    
    # Combined dimensions
    if(verbosity >= 2){
      cat(
        strftime(Sys.time(), format="%Y-%m-%d %H:%M:%S"),
        "- Combining dimensions for ATAC and RNA\n"
        )
    }
    
    COMB_LSI_name <- paste0("LSI_Combined_", i, "_ResRNA_", gsub("\\.", "-", as.character(res_RNA)), "_ResATAC_", gsub("\\.", "-", as.character(res_ATAC)), "_Iters_", iters, "_Dims_1to", dims)
    
    if (!writeLogs) {
      sink(logFile, append=TRUE)
    }
    
    proj <- addCombinedDims(
      proj, 
      reducedDims = c(
        ATAC_LSI_name,
        RNA_LSI_name
        ), 
      name = COMB_LSI_name
    )
    
    if (!writeLogs) {
      sink()
    }
    
    # UMAP and clustering for each dimensionality reduction
    if(verbosity >= 2){
      cat(strftime(Sys.time(), format="%Y-%m-%d %H:%M:%S"), "- Applying UMAP and clustering \n")
    }
    
    if (!writeLogs) {
      sink(logFile, append=TRUE)
    }
    
    proj <- addUMAP(
      proj,
      reducedDims = ATAC_LSI_name,
      name = paste0("UMAP_ATAC_", i),
      minDist = 0.8,
      force = TRUE
    )
    
    proj <- addUMAP(
      proj,
      reducedDims = RNA_LSI_name,
      name = paste0("UMAP_RNA_", i),
      minDist = 0.8,
      force = TRUE
    )
    
    proj <- addUMAP(
      proj,
      reducedDims = COMB_LSI_name,
      name = paste0("UMAP_Combined_", i),
      minDist = 0.8,
      force = TRUE
    )
    
    if (!writeLogs) {
      sink()
    }
    
    cluster_name <- paste0("Cluster_", i, "_ResRNA_", gsub("\\.", "-", as.character(res_RNA)), "_ResATAC_", gsub("\\.", "-", as.character(res_ATAC)), "_Iters_", iters, "_Dims_1to", dims)
    
    if(verbosity >= 2){
      cat(strftime(Sys.time(), format="%Y-%m-%d %H:%M:%S"), "- Saving clusters with name:", cluster_name, "\n")
    }
    
    if (!writeLogs) {
      sink(logFile, append = TRUE)
    }

    # TODO: check assumption of resolution here
    proj <- addClusters(
      proj,
      reducedDims = COMB_LSI_name,
      name = cluster_name,
      resolution = (res_ATAC + res_RNA)/2,
      force = TRUE
    )
    
    if (!writeLogs) {
      sink()
    }
    
    if(verbosity >= 1){
      cat(strftime(Sys.time(), format="%Y-%m-%d %H:%M:%S"), "- Saving cluster assignments to CSV\n")
    }
    
    currentClusteringLabels <- data.frame(proj@cellColData@listData[[cluster_name]])
    
    colnames(currentClusteringLabels) <- c(cluster_name)
    
    # View(currentClusteringLabels)
    
    appendToCSV(currentClusteringLabels, "ClusterAssignments.csv")
    
    if(verbosity >= 1){
      cat(strftime(Sys.time(), format="%Y-%m-%d %H:%M:%S"), "- Finished iteration", i, "\n")
    }
  }

  if(verbosity >= 1) {
    difference <- Sys.time() - start_time
    cat(strftime(Sys.time(), format="%Y-%m-%d %H:%M:%S"), "- Finished Loops, total time: ", elapsedTime(difference), "\n")
  }
  
  return(proj)
}
```

```{r}
proj <- testClusters(
  proj,
  resolutions_ATAC = c(0.2, 0.3), 
  resolutions_RNA = c(0.2, 0.3), 
  iterationsList = c(5, 5), 
  dimsList = c(10, 15),
  verbosity = 2
)
```

# Dune & RandIndex

```{r}
merger <- read.csv("ClusterAssignments.csv", row.names=1)

### ALL Plotted Together ###
# plot ARIs as barplot
getARIs <- ARIs(merger)

sumdata=data.frame(value=apply(getARIs,2,mean))
sumdata$key=rownames(sumdata)

# save plots
pdf(file = "RandIndex_heatmap_all.pdf", height = 15, width = 15)
plotARIs(clusMat = merger) + RotatedAxis()
dev.off()

pdf(file = "RandIndex_barplot_all.pdf", height = 8, width = 8)
ggplot(data=sumdata, aes(x=key, y=value, fill=key)) +
  geom_bar(colour="black", stat="identity") + coord_flip() + 
  theme(legend.position="none")
dev.off()
```
